WordPress/openverse
# Add API routes and controllers for additional search views

## Fixes



Fixes #2849 by @obulat
Fixes #2850 by @obulat

## Note
This PR was changed a lot after reviews, mainly removing the query parameters, so I wrote a new description in [this comment](https://github.com/WordPress/openverse/pull/2853#issuecomment-1747352637). The old description is left here for the record.

## Description


This PR adds the endpoints for additional search views for both audio and image. The endpoints start with `/v1/images` or `/v1/audio`:
- **Tag** view: `/tag/tag_name`
- **Source** view: `/source/source_name`
- **Creator** view: `/source/source_name/creator/creator_name`.
These endpoints return paginated results because they accept `page` query parameter. They also support query parameters such as `license`, `license_type`, `category`,  [image-only] `aspect_ratio`, `size`, [audio-only] `length`, `peaks`. Currently, the frontend plans and designs do not allow for querying the collections using these filter parameters, but we can add this feature later.

To enable this feature, I had to refactor the media serializers and the search controller.

### Search controller
 I extracted functions that are common for both the collection and search requests.
Both the `collection`s and the `search` endpoints now call the `search_controller.query_media` function. `query_media` builds the relevant query (collection or search), sets up other search parameters, executes the search, tallies the results and builds the search context.
The new `build_collection_query` creates a query from a dictionary of the possible query predicates (`filter`, `must_not`, `should` and `must` dictionaries). All of these dictionaries are then combined with a `Bool` query. This makes the query much less nested than the query that we currently create in the `search` method in the `search_controller`.

### Serializers
This PR adds a `MediaCollectionRequestSerializer` to allow for filtering the collections using the filters such as `license` or `category`. This would allow us to make large collections easier to browse by various criteria such as the license or category.

This new serializer reuses some common functionality from the `MediaSearchRequestSerializer`. To make it possible, I had to extract the common functionality into `MediaListRequestSerializer`.

### Sorting of the results
The results are sorted by `created_on`. The tag results also add the boost for `unstable__authority` (for the authorized requests).

## Testing Instructions

Run the app using `just up`.
Try various collection routes:
1. http://localhost:50280/v1/images/source/stocksnap (source collection, only the sources that are present in the sample data can be used to test locally)
1. http://localhost:50280/v1/images/source/flickr/creator/Manzabar/ (creator collection)
1. http://localhost:50280/v1/images/tag/cat (tag collection, images)
1. http://localhost:50280/v1/audio/tag/birds (tag collection, audio)
1. http://localhost:50280/v1/audio/tag/birds/?extension=flac (additional query parameter)
1. http://localhost:50280/v1/images/source/flickr/?page=2 (second page of results)

Try the routes that should return errors:
1. http://localhost:50280/v1/images/source/met (We don't have anything from `met` in the sample data, so the source validator throws an error
1. http://localhost:50280/v1/images/source/flickr,stocksnap (The path parameters are not split by a comma, so only a single parameter is allowed)

If there are no items corresponding to the requested values, the empty collection can be returned:
http://localhost:50280/v1/images/source/flickr/creator/Manzaba/ (the creator name is missing the last `r` letter)

Since the image and audio search had to be modified, too, check that they work as expected (the CI should check that, too).



## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
